test/libai/common/MatrixTest.java

```diff
@@ -0,0 +1,788 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ * 
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ * 
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+package libai.common;
+
+import java.util.Random;
+import libai.common.functions.Function;
+import org.junit.Test;
+
+import static org.junit.Assert.*;
+
+/**
+ *
+ * @author Federico Vera {@literal <dktcoding [at] gmail>}
+ */
+public class MatrixTest {
+    private static final double DELTA = 1e-12;
+    
+    public MatrixTest() {
+    }
+    
+    @Test
+    public void testConstructorNotIdentity1() {
+        Matrix m = new Matrix(5, 5, false);
+        for (int i = 0; i < m.getRows(); i++) {
+            assertArrayEquals(new double[]{0,0,0,0,0}, m.getRow(i), DELTA);
+        }
+    }
+    
+    @Test
+    public void testConstructorNotIdentity2() {
+        Matrix m = new Matrix(15, 5, false);
+        for (int i = 0; i < m.getRows(); i++) {
+            for (int j = 0; j < m.getColumns(); j++) {
+                assertEquals(0, m.position(i, j), DELTA);
+            }
+        }
+    }
+    
+    @Test
+    public void testConstructorIdentity1() {
+        Matrix m = new Matrix(5, 5, true);
+        for (int i = 0; i < m.getRows(); i++) {
+            double[] test = new double[5];
+            test[i] = 1;
+            assertArrayEquals(test, m.getRow(i), DELTA);
+        }
+    }
+    
+    @Test
+    public void testConstructorIdentity2() {
+        Matrix m = new Matrix(15, 5, true);
+        for (int i = 0; i < m.getColumns(); i++) {
+            double[] test = new double[5];
+            test[i] = 1;
+            assertArrayEquals(test, m.getRow(i), DELTA);
+        }
+    }
+    
+    @Test
+    public void testConstructorIdentity3() {
+        Matrix m = new Matrix(5, 15, true);
+        for (int i = 0; i < m.getRows(); i++) {
+            double[] test = new double[15];
+            test[i] = 1;
+            assertArrayEquals(test, m.getRow(i), DELTA);
+        }
+    }
+
+    @Test
+    public void testMatrixRandom() {
+        Matrix a = Matrix.random(50, 100, true);
+        Matrix b = Matrix.random(50, 100, false);
+        assertNotEquals(a, b); //<- yes... this test could actually fail...
```

Comment:
insert a small delay, Thread.sleep(1), between creations should be enough to offset the creations.